### PR TITLE
ref(rust): Don't panic in RunTaskInThreads::poll

### PR DESCRIPTION
**Pull request by loewenheim**
This removes the panics in `RunTaskInThreads` that happened in cases of `JoinErrors` or "other" `RunTaskError`s. To this end, the error type of `poll` (and, by necessity, `join`) is changed from `InvalidMessage` to a new enum `PollError` covering exactly these three cases:

```rust
pub enum PollError {
    InvalidMessage(InvalidMessage),
    JoinError(JoinError),
    Other(Box<dyn std::error::Error>),
}
```
The logging of the latter two error cases is moved from `join` to `StreamProcessor::_run_once`.

Questions:
* ~`PollError` is IMO the natural name for the new error enum, but we already have a type of that name. This doesn't necessarily cause problems, but might be confusing. Should I change the name?~ Renamed to `StrategyError`.
* Are we logging the errors in the right place (`StreamProcessor::_run_once`)?
* ~Arguably the `PollError` enum is overcomplicated—there really is no functional difference between the `JoinError` and `Other` cases. Is there value in preserving the distinction? Otherwise I'll boil it down to just two cases.~ `JoinError` variant removed.

Fixes SNS-2594.

### CODE REVIEW

**Comment by untitaker** [comment]
> Arguably the PollError enum is overcomplicated—there really is no functional difference between the JoinError and Other cases. Is there value in preserving the distinction? Otherwise I'll boil it down to just two cases.

IMO there should only be one error for all three cases (poll, join, and thread handles)

File: rust_snuba/rust_arroyo/src/processing/mod.rs [hunk 1]
```diff
@@ -298,15 +297,26 @@ impl<TPayload: Clone + Send + Sync + 'static> StreamProcessor<TPayload> {
...
+                    None => {
+                        tracing::error!("Could not find invalid message in buffer");
+                    }
                 }
-            },
+            }
+
+            Err(strategies::PollError::JoinError(error)) => {
+                let error: &dyn std::error::Error = &error;
+                tracing::error!(error, "the thread crashed");
```
Comment (untitaker) [inline rust_snuba/rust_arroyo/src/processing/mod.rs:314]: I think we still need to make sure that the consumer shuts down, so we still need to somehow return (or panic as a temporary measure)

File: rust_snuba/rust_arroyo/src/processing/strategies/healthcheck.rs [hunk 2]
```diff
@@ -49,7 +49,7 @@ impl<TPayload, Next> ProcessingStrategy<TPayload> for HealthCheck<Next>
 where
     Next: ProcessingStrategy<TPayload> + 'static,
 {
-    fn poll(&mut self) -> Result<Option<CommitRequest>, InvalidMessage> {
+    fn poll(&mut self) -> Result<Option<CommitRequest>, PollError> {
```
Comment (untitaker) [inline rust_snuba/rust_arroyo/src/processing/strategies/healthcheck.rs:52]: that is the zero-sized type, right? if we do this then we have no way to propagate invalid messages to the dlq anymore

**Comment by untitaker** [comment]
> PollError is IMO the natural name for the new error enum, but we already have a type of that name. This doesn't necessarily cause problems, but might be confusing. Should I change the name?

how about strategy error?

**Comment by untitaker** [comment]
> Arguably the PollError enum is overcomplicated—there really is no functional difference between the JoinError and Other cases. Is there value in preserving the distinction? Otherwise I'll boil it down to just two cases.

you mean boxing tokio::JoinError as well? sure why not

File: rust_snuba/rust_arroyo/src/processing/strategies/run_task.rs [hunk 3]
```diff
@@ -5,6 +5,8 @@ use crate::processing::strategies::{
 use crate::types::Message;
 use std::time::Duration;
 
+use super::StrategyError;
```
Comment (untitaker) [inline rust_snuba/rust_arroyo/src/processing/strategies/run_task.rs:8]: i think this is the only use of `super`, can you change it to `use crate::`? I personally find it easier to comprehend, especially in diffs (and it seems github's code "intelligence" also got confused, when I click on it)

**Comment by loewenheim** [comment]
> > Arguably the PollError enum is overcomplicated—there really is no functional difference between the JoinError and Other cases. Is there value in preserving the distinction? Otherwise I'll boil it down to just two cases.
> 
> you mean boxing tokio::JoinError as well? sure why not

Not just boxing it, but also consolidating the `JoinError` and `Other` variants.

**Comment by loewenheim** [comment]
Further changes:
* Added a variant `Thread` to `RunError`. This is what `StrategyError::Other` is converted to.
* In `consumer_impl`, instead of unwrapping `processor.run()`, we now log errors and exit manually.

File: rust_snuba/rust_arroyo/src/processing/mod.rs [hunk 4]
```diff
@@ -310,7 +312,7 @@ impl<TPayload: Clone + Send + Sync + 'static> StreamProcessor<TPayload> {
             }
 
             Err(StrategyError::Other(error)) => {
-                tracing::error!(error, "the thread errored");
+                return Err(RunError::Thread(error));
```
Comment (untitaker) [inline rust_snuba/rust_arroyo/src/processing/mod.rs:315]: I think this is a misnomer, the StreamProcessor doesn't know whether there was a thread at all

Comment (loewenheim) [reply rust_snuba/rust_arroyo/src/processing/mod.rs:315]: Then maybe "the strategy errored"?

Comment (untitaker) [reply rust_snuba/rust_arroyo/src/processing/mod.rs:315]: sgtm

File: rust_snuba/rust_arroyo/src/processing/strategies/healthcheck.rs [hunk 5]
```diff
@@ -2,11 +2,11 @@ use std::path::PathBuf;
 use std::time::{Duration, SystemTime};
 
 use crate::counter;
-use crate::processing::strategies::{
-    CommitRequest, InvalidMessage, ProcessingStrategy, SubmitError,
-};
+use crate::processing::strategies::{CommitRequest, ProcessingStrategy, SubmitError};
 use crate::types::Message;
 
+use super::StrategyError;
```
Comment (untitaker) [inline rust_snuba/rust_arroyo/src/processing/strategies/healthcheck.rs:8]: can you again get rid of `super`? 😅

**Review by untitaker — APPROVED** [review]

**Pull request merged**